pkg/compactor/bucket_compactor.go

```diff
@@ -394,6 +398,14 @@ func (c *BucketCompactor) runCompactionJob(ctx context.Context, job *Job) (shoul
 			return errors.Wrapf(err, "invalid result block %s", bdir)
 		}
 
+		if newMeta.MinTime == toCompactMinTime.UnixMilli() {
```

Comment:
Since we're inside `concurrency.ForEachJob()`, at this point we may have already uploaded some compacted blocks. I would move your validation before the call to `convertCompactionResultToForEachJobs`.